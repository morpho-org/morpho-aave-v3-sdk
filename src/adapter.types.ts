import {
  FetchedStatic,
  FetchedUpdated,
  MarketConfig,
  MarketData,
  MarketMapping,
  ScaledMarketData,
  ScaledMarketSupply,
  ScaledUserMarketData,
  UserMarketData,
} from "./types";

export type MarketsConfigs = MarketMapping<FetchedStatic<MarketConfig>>;
export type ScaledMarketsData = MarketMapping<
  FetchedStatic<ScaledMarketData & ScaledMarketSupply>
>;
export type MarketsData = MarketMapping<FetchedUpdated<MarketData>>;
export type UserMarketsData = MarketMapping<FetchedUpdated<UserMarketData>>;
export type ScaledUserMarketsData = MarketMapping<
  FetchedStatic<ScaledUserMarketData>
>;
