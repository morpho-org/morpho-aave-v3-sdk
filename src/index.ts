import sdk from "./configuration";

export { MorphoAaveV3Adapter } from "./MorphoAaveV3Adapter";

export default sdk;
