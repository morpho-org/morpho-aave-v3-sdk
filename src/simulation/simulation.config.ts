import { WadRayMath } from "@morpho-labs/ethers-utils/lib/maths";

export const HF_THRESHOLD = WadRayMath.WAD;
