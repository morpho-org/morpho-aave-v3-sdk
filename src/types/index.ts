export * from "./generics";
export * from "./common";
export * from "./markets";
export * from "./user";
export * from "./tx";
