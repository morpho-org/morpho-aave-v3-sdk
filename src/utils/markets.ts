import { MarketConfig, PositionType, TransactionType } from "../types";

export function isMarketPaused(
  txType: TransactionType,
  marketConfig: MarketConfig
): boolean;
export function isMarketPaused(
  PositionType: PositionType,
  marketConfig: MarketConfig
): boolean;
export function isMarketPaused(
  type: TransactionType | PositionType,
  marketConfig: MarketConfig
) {
  switch (type) {
    case TransactionType.borrow:
      return marketConfig.isBorrowPaused;
    case TransactionType.supply:
      return marketConfig.isSupplyPaused;
    case TransactionType.supplyCollateral:
      return marketConfig.isSupplyCollateralPaused;
    case TransactionType.repay:
      return marketConfig.isRepayPaused;
    case TransactionType.withdraw:
      return marketConfig.isWithdrawPaused;
    case TransactionType.withdrawCollateral:
      return marketConfig.isWithdrawCollateralPaused;
    case PositionType.borrow:
      return marketConfig.isRepayPaused && marketConfig.isBorrowPaused;
    case PositionType.supply:
      return marketConfig.isWithdrawPaused && marketConfig.isSupplyPaused;
    case PositionType.collateral:
      return (
        marketConfig.isWithdrawCollateralPaused &&
        marketConfig.isSupplyCollateralPaused
      );
  }
}
