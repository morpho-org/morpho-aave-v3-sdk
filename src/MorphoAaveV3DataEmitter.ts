import { deepCopy } from "ethers/lib/utils";
import { Subject } from "rxjs";

import { MorphoAaveV3DataHolder } from "./MorphoAaveV3DataHolder";
import { MarketsConfigs, MarketsData, UserMarketsData } from "./adapter.types";
import { FetchedStatic, FetchedUpdated, GlobalData, UserData } from "./types";

export abstract class MorphoAaveV3DataEmitter extends MorphoAaveV3DataHolder {
  /* Subjects */
  public readonly marketsConfigs$ = new Subject<MarketsConfigs>();
  public readonly marketsData$ = new Subject<MarketsData>();
  public readonly userMarketsData$ = new Subject<UserMarketsData>();
  public readonly marketsList$ = new Subject<FetchedStatic<string[]>>();
  public readonly userData$ = new Subject<FetchedUpdated<UserData>>();
  public readonly globalData$ = new Subject<FetchedUpdated<GlobalData>>();

  /* Setters */
  protected set marketsConfigs(mc: MarketsConfigs) {
    this._marketsConfigs = mc;
    this.marketsConfigs$.next(deepCopy(mc));
  }
  protected set marketsData(md: MarketsData) {
    this._marketsData = md;
    this.marketsData$.next(deepCopy(md));
  }
  protected set userMarketsData(umd: UserMarketsData) {
    this._userMarketsData = umd;
    this.userMarketsData$.next(deepCopy(umd));
  }
  protected set marketsList(ml: FetchedStatic<string[]>) {
    this._marketsList = ml;
    this.marketsList$.next(deepCopy(ml));
  }
  protected set userData(ud: FetchedUpdated<UserData>) {
    this._userData = ud;
    this.userData$.next(deepCopy(ud));
  }
  protected set globalData(gd: FetchedUpdated<GlobalData>) {
    this._globalData = gd;
    this.globalData$.next(deepCopy(gd));
  }
}
