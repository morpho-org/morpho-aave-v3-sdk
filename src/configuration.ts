import { BigNumber } from "ethers";
import { parseUnits } from "ethers/lib/utils";

const defaultRpcHttpUrl = process.env.RPC_HTTP_URL || process.env.RPC_URL;

interface Configuration {
  readonly isProd: boolean;
  readonly rpcHttpUrl?: string;
  readonly network: number;
  readonly defaultMaxIterations: {
    readonly supply: number;
    readonly borrow: number;
  };
  readonly gasLimitPercent: BigNumber;
  readonly percentApproximation: BigNumber;
  readonly txSignature?: string;
}

class MorphoAaveV3Sdk {
  private __configuration: Configuration = {
    isProd: false,
    rpcHttpUrl: defaultRpcHttpUrl,
    network: 1,
    gasLimitPercent: parseUnits("1.1", 4),
    percentApproximation: parseUnits("0.99", 4), // 99%
    defaultMaxIterations: {
      supply: 4,
      borrow: 4,
    },
  } as const;

  public get configuration() {
    return { ...this.__configuration };
  }

  public setConfiguration = (newConfig: Partial<Configuration>) => {
    this.__configuration = { ...this.__configuration, ...newConfig };
  };
}

export default new MorphoAaveV3Sdk();
