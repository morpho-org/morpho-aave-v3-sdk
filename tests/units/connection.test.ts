import { BigNumber, constants, getDefaultProvider, Wallet } from "ethers";
import { Subscription } from "rxjs";

import { MorphoAaveV3Adapter } from "../../src/MorphoAaveV3Adapter";
import { MorphoAaveV3Simulator } from "../../src/simulation/MorphoAaveV3Simulator";
import { ErrorCode, SimulationError } from "../../src/simulation/SimulationError";
import { Operation } from "../../src/simulation/simulation.types";
import { TransactionType } from "../../src/types";
import { Underlying } from "../mocks/markets";
import { ADAPTER_MOCK } from "../mocks/mock";
import configuration from "../../src/configuration";

describe("Connection", () => {
  let adapter: MorphoAaveV3Adapter;

  beforeAll(async () => {
    configuration.setConfiguration({
      rpcHttpUrl: process.env.RPC_URL,
    });
    adapter = MorphoAaveV3Adapter.fromChain({
      _provider: getDefaultProvider("https://bad-provider.io"),
    });
  });

  afterEach(async () => {});

  it("should initialize when we connect a user", async () => {
    await expect(adapter.refreshAll("latest")).rejects.toThrowErrorMatchingSnapshot();

    const signer = Wallet.createRandom().connect(getDefaultProvider(process.env.RPC_URL));
    await expect(adapter.connect(signer.address, signer)).resolves.not.toThrowError();

    expect(adapter.getUserMarketsData()).not.toEqual({});
    await expect(adapter.refetchData()).resolves.not.toThrowError();
  });
});
