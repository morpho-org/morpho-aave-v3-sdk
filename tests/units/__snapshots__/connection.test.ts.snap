// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection should initialize when we connect a user 1`] = `
"Error during initialization.

Error: Error during initialisation"
`;

exports[`Simulator should initialize when we connect a user 1`] = `
"Error during initialization.

Error: Error during initialisation"
`;
